lib/utils/BaseReporter.js

```diff
@@ -63,45 +66,76 @@ class BaseReporter extends events.EventEmitter {
         })
 
         this.on('runner:start', (runner) => {
-            if (!this.stats.runner[runner.cid]) {
-                this.stats.runner[runner.cid] = {
-                    start: new Date(),
-                    capabilities: runner.capabilities,
-                    config: runner.config,
-                    tests: []
-                }
-            }
+            this.getSpecStats(runner)
```

Comment:
> It seems like a responsibility of the BaseReporter to be able to summarise everything.

:+1: 

> The reason getTestStats() (and others) are horrible methods is that they have two purposes

Yes!

> I suppose memory is also a concern here - keeping so much test output might not work for long-running tests.

That's a valid point! Though I think that is a different topic and we should not block ourself due to this issue. I have no experiences where the critical point here is. I don't think that we store that much data and I guess we are more constraint of the CPU power (I tried to run ~50 capabilties at once and my computer started lacking).

Performance optimisations is definitely something we should care about but not for this release. I would say even after we improved mobile in WebdriverIO. Maybe [node-cache](https://www.npmjs.com/package/node-cache) could be a solution here. But before we don't have any benchmarks we shouldn't make early assumptions.

Apparently this is more complex then we initially thought. I think we should split this into smaller pieces. That said what do you think about creating classes that only purpose is to take an object of data and format it into the right format so it can be propagated to the other reporters. So maybe have `TestResult` class for the `state = 'pass'` case and a `TestState` for the `test:start` one. It doesn't matter if they only have one or two methods. The goal is to simplify the BaseReporter routines to make the code easy to read.